common/viperutil/config_util.go

```diff
@@ -109,163 +109,155 @@ func unmarshalJSON(val interface{}) (map[string]string, bool) {
+			if err != nil {
+				return data, err
+			}
+			return string(bytes), nil
+		case ok:
+			// fileName was nil
+			return nil, fmt.Errorf("Value of File: was nil")
 		}
-		v := reflect.ValueOf(data)
-		switch v.Kind() {
-		case reflect.String:
-			return data, nil
-		case reflect.Map:
-			d := data.(map[string]interface{})
-			fileName, ok := d["File"]
+	}
+	return data, nil
+}
+
+func pemBlocksFromFileDecodeHook(f reflect.Kind, t reflect.Kind, data interface{}) (interface{}, error) {
+	// "to" type should be string
+	if t != reflect.Slice {
+		return data, nil
+	}
+	// "from" type should be map
+	if f != reflect.Map {
+		return data, nil
+	}
+	v := reflect.ValueOf(data)
+	switch v.Kind() {
+	case reflect.String:
+		return data, nil
+	case reflect.Map:
+		var fileName string
+		var ok bool
+		switch d := data.(type) {
+		case map[string]string:
+			fileName, ok = d["File"]
 			if !ok {
 				fileName, ok = d["file"]
 			}
-			switch {
-			case ok && fileName != nil:
-				bytes, err := ioutil.ReadFile(fileName.(string))
-				if err != nil {
-					return data, err
-				}
-				return string(bytes), nil
-			case ok:
-				// fileName was nil
-				return nil, fmt.Errorf("Value of File: was nil")
+		case map[string]interface{}:
+			var fileI interface{}
+			fileI, ok = d["File"]
+			if !ok {
+				fileI = d["file"]
 			}
+			fileName, ok = fileI.(string)
 		}
-		return data, nil
-	}
-}
 
-func pemBlocksFromFileDecodeHook() mapstructure.DecodeHookFunc {
-	return func(f reflect.Kind, t reflect.Kind, data interface{}) (interface{}, error) {
-		// "to" type should be string
-		if t != reflect.Slice {
-			return data, nil
-		}
-		// "from" type should be map
-		if f != reflect.Map {
-			return data, nil
-		}
-		v := reflect.ValueOf(data)
-		switch v.Kind() {
-		case reflect.String:
-			return data, nil
-		case reflect.Map:
-			var fileName string
-			var ok bool
-			switch d := data.(type) {
-			case map[string]string:
-				fileName, ok = d["File"]
-				if !ok {
-					fileName, ok = d["file"]
-				}
-			case map[string]interface{}:
-				var fileI interface{}
-				fileI, ok = d["File"]
-				if !ok {
-					fileI = d["file"]
-				}
-				fileName, ok = fileI.(string)
+		switch {
+		case ok && fileName != "":
+			var result []string
+			bytes, err := ioutil.ReadFile(fileName)
+			if err != nil {
+				return data, err
 			}
-
-			switch {
-			case ok && fileName != "":
-				var result []string
-				bytes, err := ioutil.ReadFile(fileName)
-				if err != nil {
-					return data, err
+			for len(bytes) > 0 {
+				var block *pem.Block
+				block, bytes = pem.Decode(bytes)
+				if block == nil {
+					break
 				}
-				for len(bytes) > 0 {
-					var block *pem.Block
-					block, bytes = pem.Decode(bytes)
-					if block == nil {
-						break
-					}
-					if block.Type != "CERTIFICATE" || len(block.Headers) != 0 {
-						continue
-					}
-					result = append(result, string(pem.EncodeToMemory(block)))
+				if block.Type != "CERTIFICATE" || len(block.Headers) != 0 {
```

Comment:
I know this is like the original form, but - why do we really enforce a CERTIFICATE header here? 
This is a generic viper util, so what if we have here a public key or a private key ? Then the header is different. 

Also, i don't think that PEM parsing cares about header type at all - it only cares about -- BEGIN and -- END prefixes and ---- suffixes.

Reply:
Removing all of 'viperutil' is on our TODO list.  I think we'd be better off without the magic (but not quite correct or consistent magic) around things like PEM blocks.